Fix inf/nan checks in unary operations

Aswinmcw opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/27993
https://github.com/tenstorrent/tt-metal/issues/27994

### Problem description
+/-inf, nan check is not working properly

### What's changed
Fix the issue by using exponent and mantissa extraction

exponent will be 128 for +/- inf and mantissa will be 0, if !=0, then it is nan

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

Aswinmcw committed:
Fix inf/nan checks in unary operations

This commit introduces a new test function `test_inf_nan_check` to validate the behavior of various tensor operations (isinf, isnan, isposinf, isneginf) against expected outputs for special floating-point values (inf, -inf, nan). The tests cover multiple data types including float32 and bfloat16, ensuring comprehensive coverage for edge cases.

Additionally, the SFPU microcode for handling these operations has been updated to improve accuracy and performance by utilizing exponent and mantissa extraction methods.

### Changes:
- Added `test_inf_nan_check` in `test_unary.py`
- Updated SFPU microcode in `ckernel_sfpu_isinf_isnan.h` for better handling of inf/nan cases

### Checklist
- [x] New tests provide coverage for changes
- [x] CI passes

Aswinmcw committed:
Refactor SFPU microcode for isinf/isnan checks

This commit removes unnecessary variable assignments for exponent and mantissa conversions in the SFPU microcode for the isinf and isnan checks. The changes streamline the calculation process, enhancing performance without altering the functionality.

### Changes:
- Removed `outexp` and `outman` variable assignments in `ckernel_sfpu_isinf_isnan.h` for both Blackhole and Wormhole B0 implementations.

### Checklist
- [x] Code review completed
- [x] CI passes

Aswinmcw committed:
Enhance inf/nan checks in unary operation tests

This commit updates the `test_inf_nan_check` function in `test_unary.py` to include additional special floating-point values, improving the coverage of edge cases for tensor operations. The new input data now includes a broader range of values such as -5.0, 0.0, -0.0, 1e38, and 1e-45, ensuring more comprehensive validation of the behavior of operations like isinf and isnan.

### Changes:
- Expanded input tensor in `test_inf_nan_check` to include more special values.

### Checklist
- [x] New tests provide coverage for changes
- [x] CI passes

VirdhatchaniKN commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isinf_isnan.h:
```diff
@@ -35,12 +35,13 @@ template <bool APPROXIMATION_MODE, int ITERATIONS>
+        vFloat in = dst_reg[0];
+        sfpi::vInt exp = sfpi::exexp(in);
+        sfpi::vInt man = sfpi::exman9(in);
+        vFloat out = 0.0f;
```
sfpi::vConst0 ?

VirdhatchaniKN commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isinf_isnan.h:
```diff
@@ -35,12 +35,13 @@ template <bool APPROXIMATION_MODE, int ITERATIONS>
+        sfpi::vInt exp = sfpi::exexp(in);
+        sfpi::vInt man = sfpi::exman9(in);
+        vFloat out = 0.0f;
+        v_if(exp == 128 && man == 0) { out = 1.0f; }
```
sfpi::vConst1

VirdhatchaniKN commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isinf_isnan.h:
```diff
@@ -49,11 +50,13 @@ template <bool APPROXIMATION_MODE, int ITERATIONS>
+        sfpi::vInt exp = sfpi::exexp(in);
+        sfpi::vInt man = sfpi::exman9(in);
+        vFloat out = 0.0f;
+        v_if(in > 0 && exp == 128 && man == 0) { out = 1.0f; }
```
sfpi::vConst1 ?

VirdhatchaniKN commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isinf_isnan.h:
```diff
@@ -49,11 +50,13 @@ template <bool APPROXIMATION_MODE, int ITERATIONS>
+        vFloat in = dst_reg[0];
+        sfpi::vInt exp = sfpi::exexp(in);
+        sfpi::vInt man = sfpi::exman9(in);
+        vFloat out = 0.0f;
```
sfpi::vConst0 ?

VirdhatchaniKN commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isinf_isnan.h:
```diff
@@ -62,11 +65,13 @@ template <bool APPROXIMATION_MODE, int ITERATIONS>
+        vFloat in = dst_reg[0];
+        sfpi::vInt exp = sfpi::exexp(in);
+        sfpi::vInt man = sfpi::exman9(in);
+        vFloat out = 0.0f;
```
Same here and for WH as well

VirdhatchaniKN commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isinf_isnan.h:
```diff
@@ -62,11 +65,13 @@ template <bool APPROXIMATION_MODE, int ITERATIONS>
+        sfpi::vInt exp = sfpi::exexp(in);
+        sfpi::vInt man = sfpi::exman9(in);
+        vFloat out = 0.0f;
+        v_if(in < 0 && exp == 128 && man == 0) { out = 1.0f; }
```
Same here and for WH as well

Aswinmcw committed:
Update SFPU microcode for isinf/isnan checks to use constant values

This commit modifies the SFPU microcode in `ckernel_sfpu_isinf_isnan.h` for both Blackhole and Wormhole B0 implementations. The changes replace the initialization of the output variable from `0.0f` to `sfpi::vConst0`, and the assignment of `1.0f` to `sfpi::vConst1` for various floating-point checks (isinf, isposinf, isneginf, isnan). This enhances code clarity and maintains consistent handling of special floating-point values.

### Changes:
- Updated output variable initialization to use constant values for better readability and maintainability.

### Checklist
- [x] Code review completed
- [x] CI passes

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/inf_nan_fix

ncvetkovicTT commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isinf_isnan.h:
```diff
@@ -49,11 +50,13 @@ template <bool APPROXIMATION_MODE, int ITERATIONS>
+        sfpi::vInt exp = sfpi::exexp(in);
+        sfpi::vInt man = sfpi::exman9(in);
+        vFloat out = sfpi::vConst0;
+        v_if(in > 0 && exp == 128 && man == 0) { out = sfpi::vConst1; }
```
Just to check,`in>0` will check only for the sign bit? Wondering if there's an optimized way of checking this.

Aswinmcw replied:
Yes that's to check if it is + inf or - inf, is there a way to get the signbit directly? I don't find any, something like exsgnbit?

ncvetkovicTT reviewed (commented):
Seems good. Could you please add an explanation of the condition in the comment of the function? Smth like:
```
/* Checks if the sign bit of the floating point number in DEST 
is negative. Checks if the exponent is 128 and mantissa is 0. 
If all of the three conditions are met, the number is marked as 
negative infinity, so '1' is written in the location of the DEST 
where the number was stored. Otherwise, `0` is written instead 
of the number.
*/
inline void calculate_isneginf() {
```

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/inf_nan_fix

Aswinmcw committed:
Refactor isinf and isnan calculations to use signbit for improved clarity and accuracy in handling positive and negative infinity checks.

Aswinmcw committed:
Add detailed comments for isinf, isposinf, isneginf, and isnan functions to enhance code clarity and understanding of floating point checks.

ncvetkovicTT reviewed (approved):
lgtm, thanks!

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/inf_nan_fix

Aswinmcw commented:
Perf for ttnn.isposinf with tensor mixed with [float("-inf"), float("inf"), float("nan"), 5.0, -5.0, 0.0, -0.0, 1e38, 1e-45]
32x32 : 2075.00 ns
1024x1024 : 25735.00 ns
Existing non working version:
32x32 : 1892.00 ns
1024x1024 : 25929.00 ns
cc. @cmaryanTT
